File: ScaleformUI_Csharp/Menus/Pause Menus/PauseMenu/Tabs/TextTab.cs
using ScaleformUI.Elements;
using ScaleformUI.Menus;
using ScaleformUI.PauseMenus.Elements;

namespace ScaleformUI.PauseMenu
{
    public class TextTab : BaseTab
    {
        public string TextTitle { get; set; }
        public int WordWrap { get; set; }
        public string BGTextureDict { get; private set; }
        public string BGTextureName { get; private set; }
        public string RightTextureDict { get; private set; }
        public string RightTextureName { get; private set; }

        public new TextColumn LeftColumn
        {
            get => (TextColumn)base.LeftColumn;
            internal set => base.LeftColumn = value;
        }

        public TextTab(string name, string title, SColor color) : base(name, color)
        {
            TextTitle = title;
            LeftColumn = new TextColumn(0);
            _identifier = "Page_Simple";
            _type = 0;
        }

        [Obsolete("Not supported yet at the moment due to new Pause conversion")]
        public void AddTitle(string title)
        {
            if (string.IsNullOrWhiteSpace(TextTitle))
                TextTitle = title;
        }

        public void AddItem(PauseMenuItem item)
        {
            LeftColumn.AddItem(item);
        }

        public override void Populate()
        {
            for (int i = 0; i < LeftColumn.Items.Count; i++)
            {
                SetDataSlot(LeftColumn.position, i);
            }
            if(!string.IsNullOrWhiteSpace(BGTextureDict))
                Main.PauseMenu._pause.CallFunction("CALL_CUSTOM_COLUMN_FUNCTION", (int)LeftColumn.position, "SET_BACKGROUND", BGTextureDict, BGTextureName);
            if (!string.IsNullOrWhiteSpace(RightTextureDict))
                Main.PauseMenu._pause.CallFunction("CALL_CUSTOM_COLUMN_FUNCTION", (int)LeftColumn.position, "SET_RIGHT_PICTURE", RightTextureDict, RightTextureName);
        }

        public override void ShowColumns()
        {
            LeftColumn.ShowColumn();
            LeftColumn.InitColumnScroll(true, 3, ScrollType.UP_DOWN, ScrollArrowsPosition.CENTER);
            LeftColumn.SetColumnScroll(-1, -1, -1, "", LeftColumn.Items.Count < LeftColumn.VisibleItems);
            LeftColumn.HighlightColumn(true, false, true);
        }

        public override void Focus()
        {
            base.Focus();
            LeftColumn.HighlightColumn(true, false, true);
        }

        public override void SetDataSlot(PM_COLUMNS slot, int index)
        {
            LeftColumn.SetDataSlot(index);
        }

        /// <summary>
        /// Image suggested not bigger than 1280x720
        /// </summary>
        /// <param name="txd"></param>
        /// <param name="txn"></param>
        public void UpdateBackground(string txd, string txn)
        {
            BGTextureDict = txd;
            BGTextureName = txn;
            if (Parent != null && Parent.Visible)
            {
                Parent._pause._pause.CallFunction("CALL_CUSTOM_COLUMN_FUNCTION", (int)LeftColumn.position, "SET_BACKGROUND", txd, txn);
            }
        }

        /// <summary>
        /// Image suggested has size 288 X 430
        /// </summary>
        /// <param name="txd"></param>
        /// <param name="txn"></param>
        public void AddPicture(string txd, string txn)
        {
            RightTextureDict = txd;
            RightTextureName = txn;
            if (Parent != null && Parent.Visible)
            {
                Parent._pause._pause.CallFunction("CALL_CUSTOM_COLUMN_FUNCTION", (int)LeftColumn.position, "SET_RIGHT_PICTURE", txd, txn);
            }
        }
    }
}

File: ScaleformUI_Lua/src/Menus/PauseMenus/PauseMenu/Tabs/TextTab.lua
TextTab = {}
TextTab.__index = TextTab
setmetatable(TextTab, { __index = BaseTab })
TextTab.__call = function() return "TextTab" end

---@class TextTab

function TextTab.New(name, _title, color)
    local base = BaseTab.New(name, color)
    base._identifier = "Page_Simple"
    base.TextTitle = _title or "" -- unused atm because of new pause implementation
    base.BGTextureDict = ""
    base.BGTextureName = ""
    base.RightTextureDict = ""
    base.RightTextureName = ""
    base.LeftColumn = TextColumn.New(0)
    local meta = setmetatable(base, TextTab)
    meta.LeftColumn.Parent = meta
    return meta
end

function TextTab:AddTitle(title)
    if not title:IsNullOrEmpty() then
        self.TextTitle = title
    end
end

function TextTab:AddItem(item)
    self.LeftColumn:AddItem(item)
end

function TextTab:Populate()
    for i = 1, #self.LeftColumn.Items, 1 do
        self:SetDataSlot(self.LeftColumn.position, i)
    end
    if not self.BGTextureDict:IsNullOrEmpty() then
        ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("CALL_CUSTOM_COLUMN_FUNCTION", self.LeftColumn.position,
            "SET_BACKGROUND", self.BGTextureDict, self.BGTextureName)
    end
    if not self.RightTextureDict:IsNullOrEmpty() then
        ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("CALL_CUSTOM_COLUMN_FUNCTION", self.LeftColumn.position,
            "SET_RIGHT_PICTURE", self.RightTextureDict, self.RightTextureName)
    end
end

function TextTab:ShowColumns()
    self.LeftColumn:ShowColumn()
    self.LeftColumn:InitColumnScroll(true, 3, ScrollType.UP_DOWN, ScrollArrowsPosition.CENTER)
    self.LeftColumn:SetColumnScroll(-1, -1, -1, "", #self.LeftColumn.Items < self.LeftColumn.VisibleItems)
    self.LeftColumn:HighlightColumn(true, false, true)
end

function TextTab:Focus()
    BaseTab.Focus(self)
    self.LeftColumn:HighlightColumn(true, false, true)
end

function TextTab:UpdateBackground(txd, txn)
    self.BGTextureDict = txd
    self.BGTextureName = txn
    if self.Parent ~= nil and self.Parent:Visible() and self.Parent:CurrentTab() == self then
        ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("CALL_CUSTOM_COLUMN_FUNCTION", self.LeftColumn.position,
            "SET_BACKGROUND", txd, txn)
    end
end

function TextTab:SetDataSlot(slot, index)
    self.LeftColumn:SetDataSlot(index)
end

function TextTab:AddPicture(txd, txn)
    self.RightTextureDict = txd
    self.RightTextureName = txn
    if self.Parent ~= nil and self.Parent:Visible() and self.Parent:CurrentTab() == self then
        ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("CALL_CUSTOM_COLUMN_FUNCTION", self.LeftColumn.position,
            "SET_RIGHT_PICTURE", txd, txn)
    end
end